File: Lander/core/src/com/nopalsoft/lander/game/Colisiones.java
package com.nopalsoft.lander.game;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.Contact;
import com.badlogic.gdx.physics.box2d.ContactImpulse;
import com.badlogic.gdx.physics.box2d.ContactListener;
import com.badlogic.gdx.physics.box2d.Fixture;
import com.badlogic.gdx.physics.box2d.Manifold;
import com.nopalsoft.lander.game.objetos.Bomba;
import com.nopalsoft.lander.game.objetos.Estrella;
import com.nopalsoft.lander.game.objetos.Gas;
import com.nopalsoft.lander.game.objetos.Laser;
import com.nopalsoft.lander.game.objetos.Nave;
import com.nopalsoft.lander.game.objetos.Plataforma;

public class Colisiones implements ContactListener {

    WorldGame oWorld;

    public Colisiones(WorldGame oWorld) {
        this.oWorld = oWorld;
    }

    @Override
    public void beginContact(Contact contact) {

        if (contact.getFixtureA().getBody().getUserData() instanceof Nave)
            beginContactNaveOtraCosa(contact.getFixtureA(), contact.getFixtureB());
        else if (contact.getFixtureB().getBody().getUserData() instanceof Nave)
            beginContactNaveOtraCosa(contact.getFixtureB(), contact.getFixtureA());

        else if (contact.getFixtureB().getBody().getUserData() instanceof Bomba)
            beginContactBombaOtraCosa(contact.getFixtureB());

        else if (contact.getFixtureA().getBody().getUserData() instanceof Bomba)
            beginContactBombaOtraCosa(contact.getFixtureA());
    }

    public void beginContactNaveOtraCosa(Fixture nave, Fixture otraCosa) {
        Body bodyNave = nave.getBody();
        Nave oNave = (Nave) bodyNave.getUserData();

        Body bodyOtraCosa = otraCosa.getBody();
        Object oOtraCosa = bodyOtraCosa.getUserData();

        if (oOtraCosa instanceof Gas) {
            Gas obj = (Gas) oOtraCosa;

            if (obj.state == Gas.STATE_NORMAL) {
                oNave.gas += 100;
                obj.state = Gas.STATE_TOMADA;
            }
            return;
        } else if (oOtraCosa instanceof Estrella) {
            Estrella obj = (Estrella) oOtraCosa;

            if (obj.state == Estrella.STATE_NORMAL) {
                oWorld.estrellasTomadas++;
                obj.state = Estrella.STATE_TOMADA;
            }
            return;
        } else if (oOtraCosa instanceof Bomba) {
            Bomba obj = (Bomba) oOtraCosa;
            if (obj.state == Bomba.STATE_NORMAL) {
                obj.state = Bomba.STATE_TOMADA;
                oNave.getHurtByBomb(15);
                Vector2 blastDirection = bodyNave.getWorldCenter().sub(bodyOtraCosa.getWorldCenter());
                blastDirection.nor();
                bodyNave.applyLinearImpulse(blastDirection.scl(.1f), bodyNave.getWorldCenter(), true);
            }
            return;
        } else if (oOtraCosa instanceof Laser) {
            Laser obj = (Laser) oOtraCosa;
            obj.isTouchingShip = true;
            return;
        }

        float velocidadImpacto = Math.abs(bodyNave.getLinearVelocity().x) + Math.abs(bodyNave.getLinearVelocity().y);
        if (velocidadImpacto > 1.5f) {
            oNave.colision(velocidadImpacto * 2.5f);
        }
        Gdx.app.log("Velocidad Impacto", velocidadImpacto + "");

        if (oOtraCosa instanceof Plataforma)
            if (((Plataforma) oOtraCosa).isFinal)
                oNave.isLanded = true;
    }

    private void beginContactBombaOtraCosa(Fixture bomba) {
        Body bodyBomba = bomba.getBody();
        Bomba oBomba = (Bomba) bodyBomba.getUserData();
        oBomba.cambioDireccion();
    }

    @Override
    public void endContact(Contact contact) {
        Fixture a = contact.getFixtureA();
        Fixture b = contact.getFixtureB();

        if (a != null && a.getBody().getUserData() instanceof Nave)
            endContactNaveOtraCosa(contact.getFixtureA(), contact.getFixtureB());
        else if (b != null && b.getBody().getUserData() instanceof Nave)
            endContactNaveOtraCosa(contact.getFixtureB(), contact.getFixtureA());
    }

    private void endContactNaveOtraCosa(Fixture nave, Fixture otraCosa) {
        if (nave == null || otraCosa == null)
            return;

        Body bodyNave = nave.getBody();
        Nave oNave = (Nave) bodyNave.getUserData();

        Body bodyOtraCosa = otraCosa.getBody();
        Object oOtraCosa = bodyOtraCosa.getUserData();

        if (oOtraCosa instanceof Laser) {
            Laser obj = (Laser) oOtraCosa;
            obj.isTouchingShip = false;
        } else if (oOtraCosa instanceof Plataforma) {
            if (((Plataforma) oOtraCosa).isFinal)
                oNave.isLanded = false;
        }
    }

    @Override
    public void preSolve(Contact contact, Manifold oldManifold) {
    }

    @Override
    public void postSolve(Contact contact, ContactImpulse impulse) {
        if (contact.getFixtureA().getBody().getUserData() instanceof Nave)
            postSolveNaveOtraCosa(contact.getFixtureB());
        else if (contact.getFixtureB().getBody().getUserData() instanceof Nave)
            postSolveNaveOtraCosa(contact.getFixtureA());
    }

    public void postSolveNaveOtraCosa(Fixture otraCosa) {

        Body bodyOtraCosa = otraCosa.getBody();
        Object oOtraCosa = bodyOtraCosa.getUserData();

        if (oOtraCosa instanceof Bomba) {
            otraCosa.setSensor(true);
        }
    }
}

File: Lander/core/src/com/nopalsoft/lander/game/Colisiones.kt
package com.nopalsoft.lander.game

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.physics.box2d.Contact
import com.badlogic.gdx.physics.box2d.ContactImpulse
import com.badlogic.gdx.physics.box2d.ContactListener
import com.badlogic.gdx.physics.box2d.Fixture
import com.badlogic.gdx.physics.box2d.Manifold
import com.nopalsoft.lander.game.objetos.Bomba
import com.nopalsoft.lander.game.objetos.Estrella
import com.nopalsoft.lander.game.objetos.Gas
import com.nopalsoft.lander.game.objetos.Laser
import com.nopalsoft.lander.game.objetos.Nave
import com.nopalsoft.lander.game.objetos.Plataforma
import kotlin.math.abs

class Colisiones(var oWorld: WorldGame) : ContactListener {
    override fun beginContact(contact: Contact) {
        if (contact.fixtureA.body.userData is Nave) beginContactNaveOtraCosa(contact.fixtureA, contact.fixtureB)
        else if (contact.fixtureB.body.userData is Nave) beginContactNaveOtraCosa(contact.fixtureB, contact.fixtureA)
        else if (contact.fixtureB.body.userData is Bomba) beginContactBombaOtraCosa(contact.fixtureB)
        else if (contact.fixtureA.body.userData is Bomba) beginContactBombaOtraCosa(contact.fixtureA)
    }

    fun beginContactNaveOtraCosa(nave: Fixture, otraCosa: Fixture) {
        val bodyNave = nave.body
        val oNave = bodyNave.userData as Nave

        val bodyOtraCosa = otraCosa.body
        val oOtraCosa = bodyOtraCosa.userData

        when (oOtraCosa) {
            is Gas -> {
                val obj = oOtraCosa

                if (obj.state == Gas.STATE_NORMAL) {
                    oNave.gas += 100f
                    obj.state = Gas.STATE_TOMADA
                }
                return
            }

            is Estrella -> {
                val obj = oOtraCosa

                if (obj.state == Estrella.STATE_NORMAL) {
                    oWorld.estrellasTomadas++
                    obj.state = Estrella.STATE_TOMADA
                }
                return
            }

            is Bomba -> {
                val obj = oOtraCosa
                if (obj.state == Bomba.STATE_NORMAL) {
                    obj.state = Bomba.STATE_TOMADA
                    oNave.getHurtByBomb(15f)
                    val blastDirection = bodyNave.getWorldCenter().sub(bodyOtraCosa.getWorldCenter())
                    blastDirection.nor()
                    bodyNave.applyLinearImpulse(blastDirection.scl(.1f), bodyNave.getWorldCenter(), true)
                }
                return
            }

            is Laser -> {
                val obj = oOtraCosa
                obj.isTouchingShip = true
                return
            }

            else -> {
                val velocidadImpacto = abs(bodyNave.getLinearVelocity().x) + abs(bodyNave.getLinearVelocity().y)
                if (velocidadImpacto > 1.5f) {
                    oNave.colision(velocidadImpacto * 2.5f)
                }
                Gdx.app.log("Velocidad Impacto", velocidadImpacto.toString() + "")

                if (oOtraCosa is Plataforma) if (oOtraCosa.isFinal) oNave.isLanded = true
            }
        }
    }

    private fun beginContactBombaOtraCosa(bomba: Fixture) {
        val bodyBomba = bomba.body
        val oBomba = bodyBomba.userData as Bomba
        oBomba.cambioDireccion()
    }

    override fun endContact(contact: Contact) {
        val a = contact.fixtureA
        val b = contact.fixtureB

        if (a != null && a.body.userData is Nave) endContactNaveOtraCosa(contact.fixtureA, contact.fixtureB)
        else if (b != null && b.body.userData is Nave) endContactNaveOtraCosa(contact.fixtureB, contact.fixtureA)
    }

    private fun endContactNaveOtraCosa(nave: Fixture?, otraCosa: Fixture?) {
        if (nave == null || otraCosa == null) return

        val bodyNave = nave.body
        val oNave = bodyNave.userData as Nave

        val bodyOtraCosa = otraCosa.body
        val oOtraCosa = bodyOtraCosa.userData

        if (oOtraCosa is Laser) {
            val obj = oOtraCosa
            obj.isTouchingShip = false
        } else if (oOtraCosa is Plataforma) {
            if (oOtraCosa.isFinal) oNave.isLanded = false
        }
    }

    override fun preSolve(contact: Contact, oldManifold: Manifold) {
    }

    override fun postSolve(contact: Contact, impulse: ContactImpulse) {
        if (contact.fixtureA.body.userData is Nave) postSolveNaveOtraCosa(contact.fixtureB)
        else if (contact.fixtureB.body.userData is Nave) postSolveNaveOtraCosa(contact.fixtureA)
    }

    fun postSolveNaveOtraCosa(otraCosa: Fixture) {
        val bodyOtraCosa = otraCosa.body
        val oOtraCosa = bodyOtraCosa.userData

        if (oOtraCosa is Bomba) {
            otraCosa.isSensor = true
        }
    }
}
